tvb_contrib/tvb/contrib/cosimulation/models/generic_2d_oscillator_multiscale.py

```diff
@@ -0,0 +1,292 @@
+
+    NOTE: This regime, if I = 2.1, is called subthreshold regime.
+    Unstable oscillations appear through a subcritical Hopf bifurcation.
+
+
+    .. figure :: img/Generic2dOscillator_01_mode_0_pplane.svg
+    .. _phase-plane-Generic2D:
+        :alt: Phase plane of the generic 2D population model with (V, W)
+
+        The (:math:`V`, :math:`W`) phase-plane for the generic 2D population
+        model for default parameters. The dynamical system has an equilibrium
+        point.
+
+    .. automethod:: Generic2dOscillator.dfun
+
+    """
+
+    def _numpy_dfun(self, state_variables, coupling, local_coupling=0.0, ev=numexpr.evaluate):
+        r"""
+        The two state variables :math:`V` and :math:`W` are typically considered
+        to represent a function of the neuron's membrane potential, such as the
+        firing rate or dendritic currents, and a recovery variable, respectively.
+        If there is a time scale hierarchy, then typically :math:`V` is faster
+        than :math:`W` corresponding to a value of :math:`\tau` greater than 1.
+
+        The equations of the generic 2D population model read
+
+        .. math::
+                \dot{V} &= d \, \tau (-f V^3 + e V^2 + g V + \alpha W + \gamma I), \\
+                \dot{W} &= \dfrac{d}{\tau}\,\,(c V^2 + b V - \beta W + a),
+
+        where external currents :math:`I` provide the entry point for local,
+        long-range connectivity and stimulation.
+
+        """
+        V = state_variables[0, :]
+        W = state_variables[1, :]
+
+        #[State_variables, nodes]
+        c_0 = coupling[0, :]
+
+        tau = self.tau
+        I = self.I
+        a = self.a
+        b = self.b
+        c = self.c
+        d = self.d
+        e = self.e
+        f = self.f
+        g = self.g
+        beta = self.beta
+        alpha = self.alpha
+        gamma = self.gamma
+
+        lc_0 = local_coupling * V
+
+        # Pre-allocate the result array then instruct numexpr to use it as output.
+        # This avoids an expensive array concatenation
+        derivative = numpy.empty_like(state_variables)
+
+        ev('d * tau * (alpha * W - f * V**3 + e * V**2 + g * V + gamma * I + gamma *c_0 + lc_0)', out=derivative[0])
+        ev('d * (a + b * V + c * V**2 - beta * W) / tau', out=derivative[1])
+
+        return derivative
+
+    def dfun(self, vw, c, local_coupling=0.0):
+        lc_0 = local_coupling * vw[0, :, 0]
+        vw_ = vw.reshape(vw.shape[:-1]).T
+        c_ = c.reshape(c.shape[:-1]).T
+        deriv = _numba_dfun_g2d(vw_, c_, self.tau, self.I, self.a, self.b, self.c, self.d, self.e, self.f, self.g,
+                                self.beta, self.alpha, self.gamma, lc_0)
+        return deriv.T[..., numpy.newaxis]
+
+
+@guvectorize([(float64[:],) * 16], '(n),(m)' + ',()'*13 + '->(n)', nopython=True)
+def _numba_dfun_g2d(vw, c_0, tau, I, a, b, c, d, e, f, g, beta, alpha, gamma, lc_0, dx):
+    "Gufunc for reduced Wong-Wang model equations."
```

Comment:
looks like copy paste here, this is not the Wong-Wang model.  Why the need for a new model if it's the same as the existing one?

Reply:
I deleted this model from tvb_contrib and corrected the docstring for the original TVB model.